parser/yaml_output_functions.c

```diff
@@ -0,0 +1,313 @@
+/***********************************************************************
+ *                   GNU Lesser General Public License
+ *
+ * This file is part of the GFDL Flexible Modeling System (FMS).
+ *
+ * FMS is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or (at
+ * your option) any later version.
+ *
+ * FMS is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with FMS.  If not, see <http://www.gnu.org/licenses/>.
+ **********************************************************************/
+#ifdef use_yaml
+
+#include <stdio.h>
+#include <yaml.h>
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+
+struct fmsyamloutkeys {
+	char key1 [255];
+        char key2 [255];
+        char key3 [255];
+        char key4 [255];
+        char key5 [255];
+        char key6 [255];
+        char key7 [255];
+        char key8 [255];
+        char key9 [255];
+        char key10 [255];
+        char key11 [255];
+        char key12 [255];
+        char key13 [255];
+        char key14 [255];
+        char key15 [255];
+	char level2key [255];
+} fmsyamloutkeys;
+struct fmsyamloutvalues {
+        char val1 [255];
+        char val2 [255];
+        char val3 [255];
+        char val4 [255];
+        char val5 [255];
+        char val6 [255];
+        char val7 [255];
+        char val8 [255];
+        char val9 [255];
+        char val10 [255];
+        char val11 [255];
+        char val12 [255];
+        char val13 [255];
+        char val14 [255];
+        char val15 [255];
+} fmsyamloutvalues;
+/* \breif Writes the key/value pairs of the fmsyamloutkeys and fmsyamloutvalues structs
+ * \param emitter The libyaml emitter for this file
+ * \param event The libyaml eent pointer
+ * \param aindex The index of keys and vals that are being written currently
+ * \param keys The keys to be written
+ * \param vals The values correcponding to keys
+ */
+void write_keys_vals_yaml (yaml_emitter_t * emitter, yaml_event_t * event , int aindex, struct fmsyamloutkeys *keys, struct fmsyamloutvalues *vals){
+    if (keys[aindex].key1[0] !='\0')  {
+      yaml_scalar_event_initialize(event, NULL, (yaml_char_t *)YAML_STR_TAG,
+      (yaml_char_t *)keys[aindex].key1, strlen(keys[aindex].key1), 1, 0, YAML_PLAIN_SCALAR_STYLE);
+       if (!yaml_emitter_emit(emitter, event)) goto error;
```

Comment:
If there is a graceful way to write the error, free the memory, and close the file in a function, I agree.  This code is taken from a sample, so I am open to recommendation.

Reply:
I think it can just be copied over to a new function, just needs to get passed some stuff. The first set of gotos only prints a warning so they can probably be replaced with the print and a return. Then the rest can just use a function like:
```
void error(yamlname, event.type, emitter, yamlout)
```
with the same error section code.